spec: update cli verify spec for UX improvement (#440)
notaryproject/notation: specs/commandline/verify.md

## Description

Use `notation verify` command to verify signatures associated with the artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. Upon successful verification, the output message is printed out as follows:

> 

Reviewer: Wondering if we should also add another line, which signature was matched/verified by displaying its digest as well?

```text
Successfully verified signature for <registry>/<repository>@<digest>
```

Tags are mutable and a tag reference can point to a different artifact than that was signed referred by the same tag. If a `tag` is used to identify the OCI artifact, the output message is as follows:

```text

> Resolved artifact tag '<tag>' to digest '<digest>' before verification.

Reviewer 2: nit:
```suggestion
Resolved artifact tag `<tag>` to digest `<digest>` before verification.
```

> Warning: The resolved digest may not point to the same signed artifact, since tags are mutable.

Reviewer 3: Note: These outputs will be in the form of logs.

Successfully verified for <registry>/<repository>@<digest>

> ```

Reviewer: Should we insert the word "signature" in the text to say "Successfully verified **_signature_** for <registry>/<repository>@<digest>. This way as and when other tools like Docker client or Ratify incorporate notation experience they can just use the same output the Notation CLI client uses?

## Description

> Use `notation verify` command to verify signatures associated with the artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. Always verify the artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and a tag reference can point to a different artifact than that was verified.

Reviewer 3: That's not true.

It is nature to verify a tagged manifest. Since the verification will return the digest of the verified manifest. Customers can then pull by digest.

Here's the workflow to verify and pull an image `example.contoso.io/app:v1`.
1. `notary verify example.contoso.io/app:v1`
2. If the above verification is successful, extract the digest from the output.
3. Use the digest in the step to pull the image by `docker pull example.contoso.io/app@<digest>`.

Because of the above verification workflow, we need to make sure the output of `notation verify` is script friendly.

The text
```text
Successfully verified for <registry>/<repository>@<digest>
```
is not too bad to be extracted using `sed` or regex.

Reviewer 4: This output looks user-friendly. What's the requirement for a script-friendly output? Keep digest only?

Reviewer 3: For the script-friendly output, `notation` has the same issue as https://github.com/oras-project/oras/issues/638

Author: An issue #455  was created for implementing script-friendly output in later release

> Upon successful verifying, the output message is printed out as following:

Reviewer 4: Two small grammar improvements:
verifying => verification
as following => as follows, see [comparison](https://ell.stackexchange.com/questions/2747/as-following-vs-as-follows)

```text
Successfully verified for <registry>/<repository>@<digest>
```

> If a `tag` is used to identify the OCI artifact, the output message is as following:

Reviewer 4: Same as above

Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. The digest of the supplied artifact is returned upon successful verification. It is recommended that this digest reference be used to pull the artifact subsequently, as registry tags may be mutable, and a tag reference can point to a different artifact that what was verified.

Upon successful verifying, the output message is printed out as following:

```text

> Verify succeeded for <registry>/<repository>@<digest>.

Reviewer 5: ```suggestion
Successfully verified <registry>/<repository>@<digest>
```

```

If a `tag` is used to identify the OCI artifact, the output message is as following:

> ```text
> Warning: A tag is used to identify the artifact for verifying. Artifact tags are mutable. Use digests to uniquely identify artifacts and avoid mutability.
> Resolving artifact tag '<tag>' to digest '<digest>' before verifying.
> Verify succeeded for <registry>/<repository>@<digest>.

Reviewer 5: ```suggestion
```text
Warning: Always verify artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and a tag reference can point to a different artifact than the one verified.
Successfully verified <registry>/<repository>@<digest>
```

Author: Accept other comments but keep "Resolving artifact tag '<tag>' to digest '<digest>' before verifying." since this info is valid if user uses a tag. We still support the `tag`, but resolved to `digest` before verification.

    "trustPolicies": [
        {
            // Policy for all artifacts, from any registry location.
            "name": "wabbit-networks-images",                         // Name of the policy.
            "registryScopes": [ "localhost:5000/net-monitor" ],       // The registry artifacts to which the policy applies.
            "signatureVerification": {                                // The level of verification - strict, permissive, audit, skip.
                "level": "strict"
            },
            "trustStores": [ "ca:wabbit-networks" ],                  // The trust stores that contains the X.509 trusted roots.
            "trustedIdentities": [                                    // Identities that are trusted to sign the artifact.
                "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}

> ```

Reviewer: I will like to keep an example for "*" in the registry scope. So instead of replacing the above example, could we include two examples, one with "*" in registry scope, and another with a specific repository?

>         {

Reviewer 4: Could you change the link of [Notation Directory Structure](https://github.com/notaryproject/notation/blob/main/specs/directory.md) to https://notaryproject.dev/docs/tutorials/directory-structure/? I will create a PR to remove the out-of-dated docs in the Notation repo later.

            // Policy for all artifacts, from any registry location.
            "name": "wabbit-networks-images",  // Name of the policy.

>             "registryScopes": [ "localhost:5000/net-monitor" ],         // The registry artifacts to which the policy applies.

Reviewer 3: make sure that the indentation of comment blocks align each other.

}
```

> For a Linux user, store file `trustpolicy.json` under directory `$HOME/.config/notation/`.

Reviewer 3: ```suggestion
For a Linux user, store file `trustpolicy.json` under directory `${HOME}/.config/notation/`.
```

> For a Mac user, store file `trustpolicy.json` under directory `$HOME/Library/Application Support/notation/`.

Reviewer 3: ```suggestion
For a macOS user, store file `trustpolicy.json` under directory `${HOME}/Library/Application Support/notation/`.
```

> For a Window user, store file `trustpolicy.json` under directory `C:\Users\<username>\AppData\Roaming\notation\`.

Reviewer 3: ```suggestion
For a Window user, store file `trustpolicy.json` under directory `%USERPROFILE%\AppData\Roaming\notation\`.
```

Reviewer 3: Note: it is `$env:USERPROFILE` for PowerShell and `%USERPROFILE%` for cmd and file explorer.

Reviewer 3: Note: Not all users have their user profiles in `C` drive. They can be in `D` drive or even in other folders.

> For a Window user, store file `trustpolicy.json` under directory `C:\Users\<username>\AppData\Roaming\notation\`.

Reviewer 2: Do you mean "For a Windows user"? 
```suggestion
For a Windows user, store file `trustpolicy.json` under directory `%USERPROFILE%\AppData\Roaming\notation\`.
```

Example values on trust policy properties:

| Property name | Value | Meaning |
| --|--|--|
| name | "wabbit-networks-images" | The name of the policy is "wabbit-networks-images". |
| registryScopes | "localhost:5000/net-monitor" | The policy only applies to artifacts stored in repository `localhost:5000/net-monitor`. |
| registryScopes | "localhost:5000/net-monitor", "localhost:5000/nginx" | The policy applies to artifacts stored in two repositories: `localhost:5000/net-monitor` and `localhost:5000/nginx`. |
| registryScopes | "*" | The policy applies to all the artifacts stored in any repositories. |
| signatureVerification | "level": "strict" | Signature verification is performed at strict level, which enforces all validations: `integrity`, `authenticity`, `authentic timestamp`, `expiry` and `revocation`.|
| signatureVerification | "level": "permissive" | The permissive level enforces most validations, but will only logs failures for `revocation` and `expiry`. |

> | signatureVerification | "level": "audit" | The audit level only enforces signature `integrity` if a signature is present. Failure of all other validations are only logged. |

Reviewer 3: We don't have `log` implemented.

| signatureVerification | "level": "skip" | The skip level does not fetch signatures for artifacts and does not perform any signature verification. |
| trustStores | "ca:wabbit-networks" | Verify signatures using certificates stored in one trust store. The type of trust store is `ca` and the name is `wabbit-networks`, which are configured using `notation certificate add` command.|
| trustStores | "ca:wabbit-networks", "ca:rocket-networks" | Verify signatures using certificates stored in two trust stores. |
| trustedIdentities | "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder" | User only trusts the identity with specific subject. User can use `notation certificate show` command to get the `subject` info. |

> | trustedIdentities | "*" | User trusts any identity (signing certificate) issued by the CA(s) in trust stores. |

Reviewer 3: Can this table be formatted? otherwise, it is too hard to review and read in raw md file.

Reviewer 3: ```suggestion
| Property name         | Value                                                                                      | Meaning                                                                                                                                                                                           |
| --------------------- | ------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name                  | "wabbit-networks-images"                                                                   | The name of the policy is "wabbit-networks-images".                                                                                                                                               |
| registryScopes        | "localhost:5000/net-monitor"                                                               | The policy only applies to artifacts stored in repository `localhost:5000/net-monitor`.                                                                                                           |
| registryScopes        | "localhost:5000/net-monitor", "localhost:5000/nginx"                                       | The policy applies to artifacts stored in two repositories: `localhost:5000/net-monitor` and `localhost:5000/nginx`.                                                                              |
| registryScopes        | "*"                                                                                        | The policy applies to all the artifacts stored in any repositories.                                                                                                                               |
| signatureVerification | "level": "strict"                                                                          | Signature verification is performed at strict level, which enforces all validations: `integrity`, `authenticity`, `authentic timestamp`, `expiry` and `revocation`.                               |
| signatureVerification | "level": "permissive"                                                                      | The permissive level enforces most validations, but will only logs failures for `revocation` and `expiry`.                                                                                        |
| signatureVerification | "level": "audit"                                                                           | The audit level only enforces signature `integrity` if a signature is present. Failure of all other validations are only logged.                                                                  |
| signatureVerification | "level": "skip"                                                                            | The skip level does not fetch signatures for artifacts and does not perform any signature verification.                                                                                           |
| trustStores           | "ca:wabbit-networks"                                                                       | Verify signatures using certificates stored in one trust store. The type of trust store is `ca` and the name is `wabbit-networks`, which are configured using `notation certificate add` command. |
| trustStores           | "ca:wabbit-networks", "ca:rocket-networks"                                                 | Verify signatures using certificates stored in two trust stores.                                                                                                                                  |
| trustedIdentities     | "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder" | User only trusts the identity with specific subject. User can use `notation certificate show` command to get the `subject` info.                                                                  |
| trustedIdentities     | "*"                                                                                        | User trusts any identity (signing certificate) issued by the CA(s) in trust stores.                                                                                                               |
```

> User can configure multiple trust policies for different scenarios. See [Trust Policy Schema and properties](https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#trust-policy) for details.

Reviewer 3: Need a todo or issue to track work item "replace `main` with a tagged version".

```shell

# Prerequisites: Signatures are stored in a registry referencing the signed OCI artifact
# Configure trust store by adding a certificate file into trust store named "wabbit-network" of type "ca"
notation certificate add --type ca --store wabbit-networks wabbit-networks.crt

# Create a JSON file named "trustpolicy.json" under directory "{NOTATION_CONFIG}".

# Verify signatures on the supplied OCI artifact identified by the digest
notation verify localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

An example of output messages for a successful verification:

```text
Successfully verified for localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify signatures on an OCI artifact identified by a tag

A tag is resolved to a digest first before verification.

```shell
# Prerequisites: Signatures are stored in a registry referencing the signed OCI artifact

# Verify signatures on an OCI artifact identified by the tag
notation verify localhost:5000/net-monitor:v1
```

An example of output messages for a successful verification:

```text

> Warning: Always verify artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and a tag reference can point to a different artifact than the one verified.

Reviewer 2: I think we should reword this warning? Because it is okay to verify with a tag and pull the artifact with the digest later on.

Author: Check the latest update.

```

An example of output messages for a successful verification:

```text
Verify succeeded for localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify signatures on an OCI artifact identified by a tag

A tag is resolved to a digest first before verification.

```shell
# Prerequisites: Signatures are stored in a registry referencing the signed OCI artifact

# Verify signatures on an OCI artifact identified by the tag
$ notation verify localhost:5000/net-monitor:v1
```

An example of output messages for a successful verification:

```text
Warning: Tag is used. Always use digest to identify the reference uniquely and immutably.

Resolve tag `v1` to digest `sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9`
Verify succeeded for localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify succeeded using "--debug" flag

```shell
$ notation verify --debug localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9

> Use trust policy: <policy_name>

Reviewer 6: Would be good to print the whole policy and not only the name.

Author: Good idea. 

One comment: Will it be a problem if the policy is printed out with too many lines?

Author: Output message when debug option is enabled will be addressed by other PRs

signature verification failure
${signature-digest}: ${error message}

Warning:
signature verification failure, but only logged the error
${signature-digest}: ${error message}

Verify succeeded on signature ${signature-digest} for localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify failed for all signatures without using "--debug" flag

Trust policy and trust store are configured properly.

```shell
$ notation verify localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
Verify failed for all the 2 signature(s) associated with localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify failed for all signatures using "--debug" flag

Trust policy and trust store are configured properly.

```shell
$ notation verify --debug localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
Use trust policy: <policy_name>
Use certificate: <trust_store_type>/<trust_store_name>/<certificate_name>
Signature verification level: [strict/permissive/audit/skip]

Error:
signature verification failed

> ${signature-digest}: ${error message}

Reviewer 6: Printing just a digest is confusing. We should add text explaining that this is the *signature digest* and not some other digest. 

Also, printing the signature would be helpful for debugging purposes. I would suggest changing this (also in other places) to:
```
Error: Signature verification failed for artifact <artifact-registry/artifact-repo:digest>
Detailed error: <if the error provides enough context to be helpful>
Signature digest: <signature-digest>
Signature: <the actual signature>
```

Author: Just to clarify the `signature digest` suggested by you is the digest to identify the signature blob, not the digest to identify the signature manifest, right?

For adding the `signature` info, there are two problems:
- There could be many signatures failed, this will lead to very big output message.
- cose signature is in binary format, which will mess the console if the cose signature is printed out.

Author: Output message when debug option is enabled will be addressed by other PRs

> Error:

Author: There could be several signatures failed verification.  The 2nd error is for the 2nd signatures

> Error:

Author: Output message when debug option is enabled will be addressed by other PRs

}
EOF

# Verify signatures on the supplied OCI artifact identified by the digest
$ notation verify localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

An example of output messages for a successful verification:

```text
Verify succeeded for localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
```

### Verify signatures on an OCI artifact identified by a tag

A tag is resolved to a digest first before verification.

```shell
# Prerequisites: Signatures are stored in a registry referencing the signed OCI artifact

# Verify signatures on an OCI artifact identified by the tag
$ notation verify localhost:5000/net-monitor:v1
```

An example of output messages for a successful verification:

```text
Warning: A tag is used to identify the artifact for verifying. Artifact tags are mutable. Use digests to uniquely identify artifacts and avoid mutability.

> Resolving artifact tag `v1` to digest `sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9` before verifying.

Reviewer 2: ```suggestion
Resolved artifact tag `v1` to digest `sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9` before verifying.
```
